openmmtools/storage/iodrivers.py

```diff
@@ -0,0 +1,1741 @@
+
+        Parameters
+        ----------
+        path : string
+            Absolute path to the group as it appears on the NetCDF file.
+
+        Returns
+        -------
+        group : NetCDF Group
+            The group that path points to. Can be accessed by path through the ._groups dictionary after binding
+
+        """
+        # NetCDF4 creates the cascade of groups automatically or returns the group if already present
+        # To simplify code, the cascade of groups is not stored in this class until called
+        self._check_bind_to_file()
+        path = normalize_path(path)
+        self._groups[path] = self.ncfile.createGroup(path)
+        return self._groups[path]
+
+    def sync(self):
+        if self.ncfile is not None:
+            self.ncfile.sync()
+
+    def close_down(self):
+        if self.ncfile is not None:
+            # Ensure the netcdf file closes down
+            self.sync()
+            self.ncfile.close()
+            self.ncfile = None
+
+    def _check_bind_to_file(self):
+        """
+        Bind to and create the file if it does not already exist (depending on access_mode)
+
+        """
+        if self.ncfile is None:
+            if self.access_mode is None:
+                if os.path.isfile(self.file_name):
+                    self.ncfile = nc.Dataset(self.file_name, 'a')
+                else:
+                    self.ncfile = nc.Dataset(self.file_name, 'w')
+            else:
+                self.ncfile = nc.Dataset(self.file_name, self.access_mode)
+
+    def __init__(self, file_name, access_mode=None):
+        super(NetCDFIODriver, self).__init__(file_name, access_mode=access_mode)
+        # Initialize the file bind variable. All actions involving files
+        self.ncfile = None
+        self._groups = {}
+        # Bind all of the Type Handlers
+        self.set_codec(str, NCString)  # String
+        self.set_codec(int, NCInt)  # Int
+        self.set_codec(dict, NCDict)  # Dict
+        self.set_codec(float, NCFloat)  # Float
+        # List/tuple
+        self.set_codec(list, NCIterable)
+        self.set_codec(tuple, NCIterable)
+        self.set_codec(np.ndarray, NCArray)  # Array
+        self.set_codec(unit.Quantity, NCQuantity)  # Quantity
+        # Bind the metadata reader types based on the dtype string of each class
+        self._IOMetaDataReaders = {self._type_maps[key].dtype_string(): self._type_maps[key] for key in self._type_maps}
+        # Counter for auto-creating infinite iterable dimensions
+        self._auto_iterable_count = 0
+
+# =============================================================================
+# ABSTRACT TYPE HANDLER
+# =============================================================================
+
+
+class NCVariableTypeHandler(ABC):
+    """
+    Pointer class which provides instructions on how to handle a given nc_variable
+    """
+    def __init__(self, parent_handler, target, storage_object=None):
```

Comment:
Renamed them to `parent_driver` to remove confusion.

Yes! The storage object can either be the top level `ncfile` OR a group/subgroup within that file.